### PR TITLE
Provision AWS Batch resources within terraform

### PR DESCRIPTION
**Pull request by kenoir**
### What is this PR trying to achieve?

In order to effectively track and reproduce the resources used for AWS batch.

### Who is this change for?

This lays the groundwork for #505

### Have the following been considered/are they needed?

- [ ] Run `terraform apply`.

---

The intention here is to script creation of the function params for the boto3 batch client, then using provisioners (both create & destroy) execute creation and destruction of compute environment / job queue / job definitions via terraform.

Jobs will be submitted separately (quite possibly via a lambda).

### CODE REVIEW

File: docker/terraform_ci/Dockerfile [hunk 1]
```diff
@@ -19,6 +19,13 @@ COPY plan.sh /app/plan.sh
 COPY run.sh /app/run.sh
 COPY notify.sh /app/notify.sh
 
+COPY provisioners /app/provisioners
+
+WORKDIR /app/provisioners
+
+RUN pip-compile --output-file requirements.txt requirements.in
```
Comment (kenoir) [inline docker/terraform_ci/Dockerfile]: Remove `pip-compile`

File: docker/terraform_ci/provisioners/aws_batch_compute.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python3
+# -*- encoding: utf-8 -*-
+"""
+Handles creating or destroying an AWS Batch compute environment
+
+Usage:
+  aws_batch_compute.py [--create=<location>]
```
Comment (kenoir) [inline docker/terraform_ci/provisioners/aws_batch_compute.py]: `aws_batch_compute.py command --flag`

File: docker/terraform_ci/provisioners/aws_batch_compute.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        maxResults=1,
+    )
+
+    return response["computeEnvironments"].length > 0
+
+
+def _parse_json_file(file):
+    file_contents = open(file).read()
+
+    return json.loads(file_contents)
```
Comment (kenoir) [inline docker/terraform_ci/provisioners/aws_batch_compute.py]: `json.loads(open(file))`

File: docker/terraform_ci/provisioners/aws_batch_compute.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+
+def create_compute_environment(client, description):
+    name = description["computeEnvironmentName"]
+
+    if not _environment_exists(client, name):
+        client.create_compute_environment(**description)
```
Comment (kenoir) [inline docker/terraform_ci/provisioners/aws_batch_compute.py]: Log both cases.

File: docker/terraform_ci/provisioners/aws_batch_compute.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    args = docopt.docopt(__doc__)
+
+    description_file_location = args['--create']
+    compute_env_name = args['--destroy']
+
+    client = boto3.client('batch')
+
+    if create is not None:
+        description = _parse_json_file(description_file_location)
+        create_compute_environment(client, description)
```
Comment (kenoir) [inline docker/terraform_ci/provisioners/aws_batch_compute.py]: Check failure conditions.

File: terraform/batch_compute/batch_iam/batch_service.tf [hunk 6]
```diff
@@ -0,0 +1,86 @@
+data "aws_iam_policy_document" "batch_service" {
```
Comment (kenoir) [inline terraform/batch_compute/batch_iam/batch_service.tf]: Comment that these match the default roles available in AWS of a similar name.

Comment (alexwlchan) [inline terraform/batch_compute/batch_iam/batch_service.tf]: Still need this comment.

**Pull request merged**